Libraries/Azure.psm1

```diff
@@ -2397,23 +2397,73 @@ function Add-AzureAccountFromSecretsFile {
 
     if ( Test-Path $secretsFile ) {
         Write-LogInfo "$secretsFile found."
-        Write-LogInfo "------------------------------------------------------------------"
-        Write-LogInfo "Authenticating Azure PS session.."
         $XmlSecrets = [xml](Get-Content $secretsFile)
         $ClientID = $XmlSecrets.secrets.SubscriptionServicePrincipalClientID
         $TenantID = $XmlSecrets.secrets.SubscriptionServicePrincipalTenantID
         $Key = $XmlSecrets.secrets.SubscriptionServicePrincipalKey
-        $pass = ConvertTo-SecureString $key -AsPlainText -Force
-        $mycred = New-Object System.Management.Automation.PSCredential ($ClientID, $pass)
+        $AzureContextFilePath = $XmlSecrets.secrets.AzureContextFilePath
         $subIDSplitted = ($XmlSecrets.secrets.SubscriptionID).Split("-")
         $subIDMasked = "$($subIDSplitted[0])-xxxx-xxxx-xxxx-$($subIDSplitted[4])"
-
-        $null = Add-AzureRmAccount -ServicePrincipal -Tenant $TenantID -Credential $mycred
-        $selectedSubscription = Select-AzureRmSubscription -SubscriptionId $XmlSecrets.secrets.SubscriptionID
+        Write-LogInfo "------------------------------------------------------------------"
+        if ($ClientID -and $Key) {
+            Write-LogInfo "Authenticating Azure PS session using Service Principal..."
+            $pass = ConvertTo-SecureString $key -AsPlainText -Force
+            $mycred = New-Object System.Management.Automation.PSCredential ($ClientID, $pass)
+            $null = Add-AzureRmAccount -ServicePrincipal -Tenant $TenantID -Credential $mycred
+        } elseif ($AzureContextFilePath) {
+            Write-LogInfo "Authenticating Azure PS session using saved context file $AzureContextFilePath..."
+            $null = Import-AzureRmContext -Path $AzureContextFilePath
+        } else {
+            $ContextFiles = Get-ChildItem -Path $PWD -Recurse | `
+                Where-Object { $_.Name.EndsWith(".json") } | Select-String -Pattern "AzureCloud" | Select Path
+            $ContextFiles = $ContextFiles.Path | Get-Unique
+            if ($ContextFiles.Count -eq 0) {
+                Write-LogWarn "No Azure authentication methods were available in Secret File."
+                Write-LogWarn "No Azure authentication context files detected in $PWD."
```

Comment:
revise the log statement

Reply:
@juhlee-microsoft , Can you clarify what you want to revise?

Reply:
"File not found in $PWD; no Azure authentication"

Reply:
@juhlee-microsoft , if no context files, then still we have "Pre-authenticated" session as last resort.
Consider this situation -
1. User doesn't want to use / Can't use service principal due to any reason.
2. User doesn't want to use context file authentication.

... But still he wants to run the LISAv2 with his own credentials.

Then he can use Connect-AzureRmAccount and complete the login step there. And then he can initiate LISAv2 to run the tests.

Reply:
@iamshital  Here is what I can read in your new flow.

if there is no ContextFile at 2422, then it goes with the cashed authentication information. However, if not, it falls to 2431 else-block. It goes out of its if-statement block. (2454)

(the user does not have valid subscription or does not want to their personal subscription becasue it charges to personal sub)

And then it hits the same cmd at 2442 -> 2457, like 2427, which fails again. If the user has valid subscription in cached authentication but she/he does not have it in context file, you intend to go authenticated. However it does not change the bool variable at 2460. 

The whole control flow needs to be revised and make it simple. Also trying to get authenticated with cached information is not a good way because it can cause unwanted cost from the different subscription use.

Please review your design.

Reply:
@juhlee-microsoft ,
>if there is no ContextFile at 2422, then it goes with the cashed authentication information.

This is correct.

> trying to get authenticated with cached information is not a good way because it can cause unwanted cost from the different subscription use

This is incorrect.

The subscription under test is taken from the Secret File (Using tag SubscriptionID).
So, As long as your Secret File has correct subscription ID, we end up in either - 
1. User has permission to use that subscription.
2. User does not have permission to use that subscription.

In the first scenario, LISAv2 will continue to run as expected on selected subscription (despite the fact that, user might have multiple subscriptions enabled under his account). And in the second scenario, LISAv2 will exit and it will not try to use any other subscription for user might have access.

If you still have doubts, we can have call for this.